DeeperFunc/Deeper.swift

```diff
@@ -0,0 +1,75 @@
+//
+//  Deeper.swift
+//  DeeperFunc
+//
+//  Created by Ilya Puchka on 26/10/2017.
+//  Copyright © 2017 Ilya Puchka. All rights reserved.
+//
+
+import Foundation
+
+public enum Either<A, B> {
+    case left(A), right(B)
+}
+
+public typealias RouteComponents = (path: [String], query: [String: String])
+
+public protocol PatternState {}
```

Comment:
Following phantom types represent pattern states:
- pattern can have only path and no query, i.e. `"users" /> :int`
- pattern can have path and query, i.e. `"users" /> :int .? string("sort")`
- pattern can be "closed", meaning it does not expect any more pattern components (after `any` pattern used in the middle), i.e. ` any /> "users"` or `"users" /> any`, or any simple pattern with just path or path and query
- pattern can be "open", meaning it needs some subsequent pattern component to be closed, i.e. in `"users" /> any  /> "info"` it's the state of the patter after `any` is applied